emote/extra/onnx_storage.py

```diff
@@ -0,0 +1,164 @@
+"""
+
+"""
+import logging
+import time
+
+from abc import ABC, abstractmethod
+from queue import Empty, Queue
+from threading import Event
+from typing import Mapping, Optional, Protocol, Sequence
+
+from google.protobuf import text_format
+from onnx import ModelProto, helper
+
+from emote.utils.timed_call import BlockTimers
+
+from .crud_storage import CRUDStorage, StorageItem, StorageItemHandle
+
+
+class QueuedExport:
+    def __init__(self, metadata: Optional[Mapping[str, str]]):
+        self.metadata = metadata
+        self.return_value = None
+        self._event = Event()
+
+    def process(self, storage: "OnnxStorage"):
+        self.return_value = storage._export_onnx(self.metadata)
+        self._event.set()
+
+    def block_until_complete(self):
+        self._event.wait()
+        return self.return_value
+
+
+def _get_version():
+    try:
+        from importlib.metadata import version
+
+        return version("emote-rl")
+    except:  # noqa
+        return "unknown-version"
+
+
+def _save_protobuf(path, message, as_text: bool = False):
+    import os
+
+    dir_name = os.path.dirname(path)
+    if dir_name:
+        os.makedirs(dir_name, exist_ok=True)
+    if as_text:
+        with open(path, "w") as f:
+            f.write(text_format.MessageToString(message))
+    else:
+        with open(path, "wb") as f:
+            f.write(message.SerializeToString())
+
+
+class OnnxStorage(ABC):
+    """Handles onnx exports of a ML policy.
+
+    Call `export` whenever you want to save an onnx version of the current model.
+
+    This is a mixin class that can be used with diamond inheritance to
+    create a specialized ONNX exporter.
+
+    Parameters:
+    :param directory: path to the directory where the files should be created. If it does not exist
+                      it will be created.
+    :param prefix: all file names will have this prefix.
+    """
+
+    def __init__(self, *args, directory: str, prefix: str = "savedmodel_", **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.storage = CRUDStorage(directory, prefix, extension="onnx")
+        self.queued_exports = Queue()
+        self.export_counter = 0
+
+        # Cache the version tag on startup.
+        # It takes about 20ms to calculate
+        self.version_tag = _get_version()
+        self.scopes = BlockTimers()
+
+    @abstractmethod
+    def generate_onnx(self) -> ModelProto:
+        """Generate a ModelProto for the policy."""
+        ...
+
+    def process_pending_exports(self):
```

Comment:
This is only used in _export. Maybe inline it?

Reply:
Ah good catch. No. It has to be called from a callback somewhere :-)